app/Repositories/Base/BaseRepository.php

```diff
@@ -0,0 +1,253 @@
+<?php
+
+namespace App\Repositories\Base;
+
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Collection;
+use Exception;
+
+abstract class BaseRepository implements BaseRepositoryContract
```

Comment:
repository dùng làm gì em?

Reply:
Repository dùng để tạo 1 kênh duy nhất để liên kết Models đến DB, nhằm tái sử dụng code và tránh lặp các đoạn code dài khi viết controller (các controller sẽ y hệt nhau, chỉ khác biến truyền vào nhưng vẫn phải viết lại liên tục nếu chỉ dùng Eloquent đơn thuần). Ngoài ra thì việc đó giúp bảo trì và phát triển về sau tốt hơn, vì thay vì thay thế và thêm chức năng vào mọi controller ta chỉ cần thêm vào BaseRepo. ngoài ra thì em biết thêm rằng là nó tạo 1 lớp trung gian cho DataAccess và Business Logic, thay vì User::all() ta sẽ có $this->repo->all(), điều này em được đọc là làm tăng tính bảo mật nhưng vì sao lại tăng và tăng ở mức nào thì em chưa thực sự hiểu rõ ạ.